File: src/storage.ts
import { mkdirSync, existsSync, readFileSync, writeFileSync, appendFileSync } from 'node:fs';
import { dirname, join } from 'node:path';

export type JsonValue = null | boolean | number | string | JsonValue[] | { [key: string]: JsonValue };
export type JsonObject = { [key: string]: JsonValue };

export const SENSITIVE_KEYS = new Set([
  'token', 'access_token', 'refresh_token', 'session', 'session_id', 'cookie', 'cookies',
  'authorization', 'account_number', 'account_no', 'accountnumber', 'accountno', 'accountid',
  'account_id', 'acctno', 'acct_no', 'secret', 'password',
]);

export function utcNow(): string {
  return new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
}

export function dataDir(base = 'data'): string {
  mkdirSync(base, { recursive: true });
  return base;
}

export function quoteHistoryPath(ticker: string, base = 'data'): string {
  return join(dataDir(base), 'quotes', `${ticker.toUpperCase()}.jsonl`);
}

export function snapshotPath(name: string, base = 'data'): string {
  return join(dataDir(base), 'snapshots', `${name}.jsonl`);
}

export function watchlistPath(base = 'data'): string {
  return join(dataDir(base), 'watchlist.json');
}

export function redact<T extends JsonValue>(value: T): T | JsonValue {
  if (Array.isArray(value)) return value.map((item) => redact(item)) as JsonValue;
  if (value && typeof value === 'object') {
    const out: JsonObject = {};
    for (const [key, item] of Object.entries(value)) {
      const norm = key.toLowerCase().replaceAll('-', '_');
      out[key] = SENSITIVE_KEYS.has(norm) ? '<redacted>' : redact(item as JsonValue);
    }
    return out;
  }
  return value;
}

export function appendJsonl(path: string, record: JsonObject): void {
  mkdirSync(dirname(path), { recursive: true });
  appendFileSync(path, `${JSON.stringify(record)}\n`, 'utf8');
}

export function readJsonl(path: string): JsonObject[] {
  if (!existsSync(path)) return [];
  return readFileSync(path, 'utf8')
    .split(/\r?\n/)
    .filter((line) => line.trim().length > 0)
    .map((line) => JSON.parse(line) as JsonObject);
}

export function readWatchlist(base = 'data'): string[] {
  const path = watchlistPath(base);
  if (!existsSync(path)) return [];
  const payload = JSON.parse(readFileSync(path, 'utf8')) as unknown;
  if (!Array.isArray(payload)) return [];
  return [...new Set(payload.map(String).map((item) => item.toUpperCase()).filter(Boolean))].sort();
}

export function writeWatchlist(tickers: string[], base = 'data'): string {
  const path = watchlistPath(base);
  mkdirSync(dirname(path), { recursive: true });
  const cleaned = [...new Set(tickers.map((item) => item.toUpperCase()).filter(Boolean))].sort();
  writeFileSync(path, `${JSON.stringify(cleaned, null, 2)}\n`, 'utf8');
  return path;
}

File: quantops_cli/storage.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

SENSITIVE_KEYS = {
    "token", "access_token", "refresh_token", "session", "session_id", "cookie", "cookies",
    "authorization", "account_number", "account_no", "accountnumber", "accountno", "accountid", "account_id", "acctno", "acct_no", "secret", "password",
}


def utc_now() -> str:
    return datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def data_dir(base: str | Path | None = None) -> Path:
    root = Path(base) if base else Path.cwd() / "data"
    root.mkdir(parents=True, exist_ok=True)
    return root


def redact(value: Any) -> Any:
    if isinstance(value, dict):
        cleaned = {}
        for key, item in value.items():
            norm = str(key).lower().replace("-", "_")
            cleaned[key] = "<redacted>" if norm in SENSITIVE_KEYS else redact(item)
        return cleaned
    if isinstance(value, list):
        return [redact(item) for item in value]
    return value


def append_jsonl(path: Path, record: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    with path.open("a", encoding="utf-8") as handle:
        handle.write(json.dumps(record, ensure_ascii=False, sort_keys=True) + "\n")


def read_jsonl(path: Path) -> list[dict[str, Any]]:
    if not path.exists():
        return []
    return [json.loads(line) for line in path.read_text(encoding="utf-8").splitlines() if line.strip()]


def quote_history_path(ticker: str, base: str | Path | None = None) -> Path:
    return data_dir(base) / "quotes" / f"{ticker.upper()}.jsonl"


def snapshot_path(name: str, base: str | Path | None = None) -> Path:
    return data_dir(base) / "snapshots" / f"{name}.jsonl"


def watchlist_path(base: str | Path | None = None) -> Path:
    return data_dir(base) / "watchlist.json"


def read_watchlist(base: str | Path | None = None) -> list[str]:
    path = watchlist_path(base)
    if not path.exists():
        return []
    payload = json.loads(path.read_text(encoding="utf-8"))
    if not isinstance(payload, list):
        return []
    return sorted({str(item).strip().upper() for item in payload if str(item).strip()})


def write_watchlist(tickers: list[str], base: str | Path | None = None) -> Path:
    path = watchlist_path(base)
    path.parent.mkdir(parents=True, exist_ok=True)
    cleaned = sorted({str(item).strip().upper() for item in tickers if str(item).strip()})
    path.write_text(json.dumps(cleaned, ensure_ascii=False, indent=2) + "\n", encoding="utf-8")
    return path
